### PR TITLE
Add code coverage exclusions for ASSERT, etc.

### PR DESCRIPTION
**Pull request by prakashsurya**
This patch adds code coverage exclusions for the ASSERT and related
macros. Often, these statements will include conditionals, and all
branches of the conditionals aren't expected to be executed.

Thus, when collecting code coverage with branch data, the coverage
report will show these conditional as partially executed. While that's
true, it's also misleading, since that's actually expected.

This patch attempts to address this issue by preventing code coverage
from being generated for these lines altogether. This will prevent these
lines from being mis-interpreted as a lack of coverage, and reduce the
distraction that they can create.

Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Prakash Surya <prakash.surya@delphix.com>

### CODE REVIEW

**Comment by prakashsurya** [comment]
This is a little frustrating, but I want to document the behavior I'm seeing. So, when using this change and building code coverage reports locally, I see `ASSERT` statements properly show up as not executed:

branch-assert

In that screenshot, one can see the `ASSERT3S` call in `ztest_random` doesn't show any hits or misses, which is expected based on my change in this PR.

But when using the codecov UI, the same `ASSERT3S` call is show yellow, indicating it was a "partial hit":

codecov-assert

The automation uses `make code-coverage-capture` to generate the report that it uploads to Codecov, and that's what I used locally to generate the HTML based report, so I'm not sure what could be going wrong here.

One difference between the Codecov report, and my local one, is we upload the `.info` file to Codecov. So, perhaps that file doesn't contain the data that would tell Codecov to ignore these lines, whereas the HTML report does have that information? I'm not really sure at this point.

I would have expected both the local HTML report, and the Codecov UI to show roughly the same coverage report, but clearly that's not the case (at least for these `ASSERT` lines).

**Comment by prakashsurya** [comment]
Experimenting some more, I pushed this change to my personal repository's master branch, and then uploaded my local coverage report to Codecov. When I did this, the `ASSERT` line properly shows up as not executed (just like my local HTML report):

mycodecov-assert

See [here](https://codecov.io/gh/prakashsurya/zfs/src/8419f6fa898161a52726229c3f6af2cbbdfd1ef5/cmd/ztest/ztest.c) for a link to the codecov report where I gathered the image above.

So, I'm left wondering if this patch actually does work, it's just that the Codecov UI won't properly reflect the exclusions until the exclusions land to master? I'm not really sure.

With that said, I'm starting to lean towards landing this (or some variation on it) since it seems to be working locally, even if the Codecov UI isn't working quite like we'd expect.

@behlendorf what're your thoughts?

This also makes me wonder if we could create a tarball of the HTML coverage report and expose that via buildbot, so we could download it and view it locally. That might be useful to help make it easier to show discrepancies between what lcov generates, and what Codecov displays.

**Comment by prakashsurya** [comment]
> This also makes me wonder if we could create a tarball of the HTML coverage report and expose that via buildbot, so we could download it and view it locally.

I've went ahead and opened the following PR to try and implemented this functionality: https://github.com/zfsonlinux/zfs-buildbot/pull/112

File: lib/libspl/include/assert.h [hunk 1]
```diff
@@ -54,26 +55,29 @@ libspl_assertf(const char *file, const char *func, int line, char *format, ...)
...
+/* LCOV_EXCL_STOP */
 
 #ifdef verify
 #undef verify
 #endif
 
 #define	VERIFY(cond)							\
-	(void) ((!(cond)) &&						\
-	    libspl_assert(#cond, __FILE__, __FUNCTION__, __LINE__))
+	(void) (("LCOV_EXCL_LINE") && ((!(cond)) &&			\
```
Comment (behlendorf) [inline lib/libspl/include/assert.h]: I think it would be preferable to use `LCOV_EXCL_BR_LINE` for all these conditionals if does work as expected.

Comment (prakashsurya) [reply lib/libspl/include/assert.h]: OK.

If we plan to move forward with the buildbot PR to expose the HTML pages, then I can use that to try out both approaches; and we can compare the HTML report with Codecov for both.

If not, I'll do that same testing locally and create the HTML reports for both, so we can evaluate the results of both approaches.

Comment (behlendorf) [reply lib/libspl/include/assert.h]: Let's go forward with the buildbot PR.

**Comment by prakashsurya** [comment]
I'm closing this PR. After some more investigation, my results above regarding the `LCOV_EXCL_LINE` working is completely false; I think I actually forgot to specify `--enable-debug`.

After trying to repeat my results, and using `--enable-debug`, I wasn't able to get any of the exclusion markers to work (`_BR_LINE`, or `_LINE`) when they were inside a macro.

Further, I started reading the `geninfo` tool's source code since it's a perl script, and it appears that this tool is looking for these exclusion macros in the `.gcov` file. The `.gcov` file is the original `.c` file (prior to applying any pre-processor) with code coverage data super-imposed. Thus, the exclusion markers will not show up in the `.gcov` file when they're inside of a C macro.

For example, if I look at `ztest.c.gcov`, I can see lines like the following:
```
       68:  920:        VERIFY(nvlist_alloc(&file, NV_UNIQUE_NAME, 0) == 0);
call    0 returned 100%
branch  1 taken 0% (fallthrough)
branch  2 taken 100%
call    3 never executed
       68:  921:        VERIFY(nvlist_add_string(file, ZPOOL_CONFIG_TYPE, VDEV_TYPE_FILE) == 0);
call    0 returned 100%
branch  1 taken 0% (fallthrough)
branch  2 taken 100%
call    3 never executed
       68:  922:        VERIFY(nvlist_add_string(file, ZPOOL_CONFIG_PATH, path) == 0);
call    0 returned 100%
branch  1 taken 0% (fallthrough)
branch  2 taken 100%
call    3 never executed
       68:  923:        VERIFY(nvlist_add_uint64(file, ZPOOL_CONFIG_ASHIFT, ashift) == 0);
call    0 returned 100%
branch  1 taken 0% (fallthrough)
branch  2 taken 100%
call    3 never executed
```
as one can see, the `LCOV_EXCL_LINE` markers aren't found in this file, and thus, `geninfo` doesn't know that it shouldn't include these lines in the coverage report.

So, to do this, we'd need to annotate the source lines where ASSERT/etc are called (i.e. not in the macro), or do some crazy/clever post-processing on the `.gcov` file to add the exclusion markers "on the fly".

**Pull request closed without merging**